pbm/pbm_util.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright (c) 2017 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pbm
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/vmware/govmomi/pbm/types"
+)
+
+// A struct to capture pbm create spec details.
+type CapabilityProfileCreateSpec struct {
+	Name           string
+	Description    string
+	Category       string
+	CapabilityList []Capability
+}
+
+// A struct to capture pbm capability instance details.
+type Capability struct {
+	ID           string
+	Namespace    string
+	PropertyList []Property
+}
+
+// A struct to capture pbm property instance details.
+type Property struct {
+	ID       string
+	Operator string
+	Value    string
+	DataType string
+}
+
+func CreatePbmCapabilityProfileSpec(pbmCreateSpec CapabilityProfileCreateSpec) (types.PbmCapabilityProfileCreateSpec, error) {
+	var pbmCapabilityProfileSpec types.PbmCapabilityProfileCreateSpec
+	capabilities, err := createCapabilityInstances(pbmCreateSpec.CapabilityList)
+	if err != nil {
+		return pbmCapabilityProfileSpec, err
+	}
+
+	pbmCapabilityProfileSpec = types.PbmCapabilityProfileCreateSpec{
+		Name:        pbmCreateSpec.Name,
+		Description: pbmCreateSpec.Description,
+		Category:    pbmCreateSpec.Category,
+		ResourceType: types.PbmProfileResourceType{
+			ResourceType: string(types.PbmProfileResourceTypeEnumSTORAGE),
+		},
+		Constraints: &types.PbmCapabilitySubProfileConstraints{
+			SubProfiles: []types.PbmCapabilitySubProfile{
+				types.PbmCapabilitySubProfile{
+					Capability: capabilities,
+				},
+			},
+		},
+	}
+	return pbmCapabilityProfileSpec, nil
+}
+
+func createCapabilityInstances(rules []Capability) ([]types.PbmCapabilityInstance, error) {
+	var capabilityInstances []types.PbmCapabilityInstance
+	for _, capabilityRule := range rules {
+		capability := types.PbmCapabilityInstance{
+			Id: types.PbmCapabilityMetadataUniqueId{
+				Namespace: capabilityRule.Namespace,
+				Id:        capabilityRule.ID,
+			},
+		}
+
+		var propertyInstances []types.PbmCapabilityPropertyInstance
+		for _, propertyRule := range capabilityRule.PropertyList {
+			property := types.PbmCapabilityPropertyInstance{
+				Id: propertyRule.ID,
+			}
+			if propertyRule.Operator != "" {
+				property.Operator = propertyRule.Operator
+			}
+			var err error
+			switch strings.ToLower(propertyRule.DataType) {
```

Comment:
Maybe you could use reflection instead of having to specify DataType, for example:
https://github.com/vmware/govmomi/blob/master/property/filter.go#L77

But I suppose that won't work well in the case of the "set" type.